src/error.rs

```diff
@@ -0,0 +1,107 @@
+use std::error::Error;
+use std::fmt;
+use std::io;
+use color::ColorType;
+use image::ImageFormat;
+
+/// An enumeration of Image errors
+#[derive(Debug)]
+pub enum ImageError {
+    /// The Image is not formatted properly
+    FormatError(String),
+
+    /// The Image's dimensions are either too small or too large
+    DimensionError,
```

Comment:
I would find it good to remove the ambiguity of `too small or too large` in the error message here. Maybe an `expected: Option<(u32, u32)>` and an `actual: (u32, u32)` field which could give much better error formatting.

Reply:
Thinking about this more, it isn't really clear to me what this error is even supposed to mean. How could an image be too small? We support 1x1 images and you can't get smaller than that! I think this variant was actually created for DXT images which error out if the width and height aren't multiples of 4 or dimensions that overflow a u32

Reply:
It seems like it is mostly used in dynamicimage and generally when the size of image data `!= width * height`. A `NotEnoughData` or something like that is maybe better?

Reply:
I don't quite undersand the usage in `dynimage.rs`. It seems to protect against a decoder reporting a size that doesn't match the pixel data returned. But in that case, shouldn't the decoder already return an error for reading? Wouldn't a mismatching success be an error in the implementation of the decoder rather than the input? I feel like an assertion could be fine here.

Opposing this is the assertion that the size fits within `u32`! That is not the problem of a buggy decoder but simply not supported for actual images. *Here* we could return `DimensionError` with good reason.

In `DXT`, the decoder will also check that the size is a multiple of 4 in all directions. It seem this limit is imposed by the format. This one would also fit `DimensionError` but neither `too small` no `too large` is correct.

Finally, in `flat` the check if a buffer is well-formed for being used as an image includes ensuring all indices are smaller than the length. That reports its own error type first as `TooLarge` which then gets converted into `ImageError::DimensionError` in the `From<_>` impl.